zim/gui/pageview/__init__.py

```diff
@@ -494,63 +521,93 @@ def image_file_load_pixels(file, width_override=-1, height_override=-1):
 		# if the file does not exist, no need to make the effort of trying to read it
 		raise FileNotFoundError(file.path)
 
-	b_size_override = width_override > 0 or height_override > 0
-	if b_size_override and (width_override <= 0 or height_override <= 0):
-		w, h = image_file_get_dimensions(file.path) # can raise
-		if height_override <= 0:
-			height_override = int(h * width_override / w)
-		else:
-			width_override = int(w * height_override / h)
-
-	# Let GTK try reading the file
+	need_switch_to_fallback = True
 	try:
 		pixbuf = GdkPixbuf.Pixbuf.new_from_file(file.path)
+	except GLib.GError:
+		logger.debug(f'GTK failed to read image, using Pillow fallback: {file.path}')
+	else:
+		need_switch_to_fallback = False
 
-		if b_size_override:
-			pixbuf = pixbuf.scale_simple(width_override, height_override, GdkPixbuf.InterpType.BILINEAR)
-				# do not use new_from_file_at_size() here due to bug in Gtk for GIF images, see issue #1563
-
-		pixbuf = rotate_pixbuf(pixbuf)
-
-	except:
-		logger.debug('GTK failed to read image, using Pillow fallback: %s', file.path)
-
-		from PIL import Image # load Pillow only if necessary
-
-		with Image.open(file.path) as img_pil:
-
-			# resize if a specific size was requested
-			if b_size_override:
-				logger.debug('PIL resizing %s %s', width_override, height_override)
-				img_pil = img_pil.resize((width_override, height_override))
+	if need_switch_to_fallback:
+		try:
+			# load Pillow only if necessary
+			# noinspection PyUnresolvedReferences
+			from PIL import Image, UnidentifiedImageError
+		except ImportError:
+			pass
+		else:
+			try:
+				with Image.open(file.path) as image:
+					pixbuf = _convert_pillow_image_to_pixbuf(image)
+			except UnidentifiedImageError:
+				logger.debug(f'Pillow failed to read image: {file.path}')
+			else:
+				need_switch_to_fallback = False
+
+	if need_switch_to_fallback:
+		error_message = f'No available fallback for load this image: {file.path}'
+		logger.debug(error_message)
+		raise TypeError(error_message)
+
+	# Let's try to find and remember the orientation before scaling,
+	# 	because we lose metadata when changing images.
+	orientation = None
+	mimetype = file.mimetype()
+	if mimetype in {'image/jpeg', 'image/tiff'}:
+		# Gtk can detect orientation in jpeg|tiff images only
+		# See docs: https://docs.gtk.org/gdk-pixbuf/method.Pixbuf.get_option.html#description
+		orientation = pixbuf.get_option('orientation')
+	if mimetype in {'image/webp', 'image/png'}:
+		# if possible, we will find orientation of the image using Pillow,
+		# 	if it is not available, we will display image it as is.
+		try:
+			# noinspection PyUnresolvedReferences
+			from PIL import Image, UnidentifiedImageError, __version__ as pillow_version_string
+		except ImportError:
+			pass
+		else:
+			pillow_version = tuple(map(int, pillow_version_string.split('.')))
+			try:
+				with Image.open(file.path) as image:
```

Comment:
Loading the image twice seems excessive, could the reading of exif data be merged to the prior read? Also, I'm not sure whether supporting EXIF orientation tags in PNGs is worth the cost of loading Pillow, given their rarity compared to `pasted_image.png`s https://stackoverflow.com/questions/9542359/does-png-contain-exif-data-like-jpg/9576717 - but I'll leave that to Jaap to decide.

Reply:
Pillow uses lazy loading  
I specifically checked 🧐  
And everything is bad 😯  
PNGs without orientation are apparently loaded completely 😭  
Unlike other options 

https://gist.github.com/Toggoren/0adfb87dab4e1c2f5b4e31e78853752c#file-gistfile1-py-L159-L167

```
f0 | mean: 0.027517 0.029340 0.937864 | median: 0.024205 0.025645 0.943845
 ^                ^        ^        ^
 |                |        |        |
 |                |        |        Just X/Y
 |                |        (Y): time required to load image into pixbuf
 |                (X): time required to load image exif data
 Orientation value. 0 means no data
```

I don't know what to do 😕

Reply:
I appreciate your thoroughness! PNG exif data just is kind of a hacky addition to the spec and there ain't two ways about that. Well - we could offer to rewrite PNGs with their embedded orientation as an extra tool :P

I'd just let PNGs retain their orientation, but if you want more [pain](https://github.com/Toggoren/PainNotes) another way would be a hidden configuration preference to enable PNG exif support - up to you.